File: app/src/main/java/org/neuverse/resource/UsuarioResource.java
package org.neuverse.resource;
import  org.neuverse.entity.*;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.*;
import java.net.URI;
import java.util.List;
import java.util.UUID;

@Path("/usuarios")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class UsuarioResource {

    @GET
    public List<Usuario> listAll() {
        return Usuario.listAll();
    }

    @GET
    @Path("/{id}")
    public Response getById(@PathParam("id") UUID id) {
        Usuario usuario = Usuario.findById(id);
        if (usuario == null) {
            throw new NotFoundException();
        }
        return Response.ok(usuario).build();
    }

    @POST
    @Transactional
    public Response create(Usuario incoming, @Context UriInfo uriInfo) {
        if (incoming == null) {
            throw new BadRequestException("Corpo da requisição ausente.");
        }
        if (incoming.email == null || incoming.email.isBlank()) {
            throw new BadRequestException("Campo 'email' é obrigatório.");
        }

        // Monta a entidade a ser persistida
        Usuario u = new Usuario();
        u.id = (incoming.id != null) ? incoming.id : UUID.randomUUID();
        u.nome = incoming.nome;
        u.email = incoming.email;
        u.onl = incoming.ativo;                // cuidado: é primitivo (boolean), default = false se não enviado
        u.cargo = incoming.cargo;
        u.cpf = incoming.cpf;
        u.dataAdmissao = incoming.dataAdmissao;
        u.criado_em = incoming.criado_em;        // opcional; se quiser confiar no default do banco, pode deixar null

        try {
            u.persist();             // Panache persist
            // garante geração de erro imediato se violar unique/email etc.
            Usuario.getEntityManager().flush();

            URI location = uriInfo.getAbsolutePathBuilder().path(u.id.toString()).build();
            return Response.created(location).entity(u).build();
        } catch (Exception e) {
            // Mapeia violação de unicidade do email para 409
            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
            if (msg.contains("unique") || msg.contains("duplicate") || msg.contains("usuarios_email")) {
                throw new WebApplicationException("Email já está em uso.", Response.Status.CONFLICT);
            }
            // outros erros (ex.: RLS bloqueando INSERT)
            throw new WebApplicationException("Falha ao inserir usuário: " + e.getMessage(),
                    Response.Status.INTERNAL_SERVER_ERROR);
        }
    }
}

File: back-end/src/main/kotlin/org/neuverse/resource/UsuarioResource.kt
package org.neuverse.resource

import jakarta.annotation.security.RolesAllowed
import jakarta.enterprise.context.RequestScoped
import jakarta.inject.Inject
import jakarta.transaction.Transactional
import jakarta.ws.rs.*
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import org.neuverse.dto.UsuarioCreateDTO
import org.neuverse.dto.UsuarioResponseDTO
import org.neuverse.entity.Usuario
import org.neuverse.repository.UsuarioRepository
import org.neuverse.service.PasswordService
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID
import org.neuverse.enums.Role

@Path("/usuarios")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@RequestScoped
class UsuarioResource @Inject constructor(
    private val repo: UsuarioRepository,
    private val passwordService: PasswordService
) {

    /**
     * Lista usuários.
     *
     * - Se `cargo` NÃO vier: lista todos.
     * - Se `cargo` vier (1=ADMIN, 2=DIRETOR, 3=FUNCIONARIO): filtra por cargo.
     *
     * Ex.: GET /usuarios?cargo=2 -> todos diretores
     */
    @GET
    fun listAll(
        @QueryParam("cargo") cargo: Role?
    ): List<UsuarioResponseDTO> {
        val usuarios: List<Usuario> =
            if (cargo == null) {
                repo.listAll()
            } else {
                // usa Panache para filtrar pelo campo cargo (int2 no banco)
                repo.list("cargo = ?1", cargo)
            }

        return usuarios.map { UsuarioResponseDTO.fromEntity(it) }
    }

    @GET
    @Path("/{id}")
    fun getOne(@PathParam("id") id: UUID): Response {
        val usuario = repo.findByIdUuid(id)
            ?: return Response.status(Response.Status.NOT_FOUND).build()

        return Response.ok(UsuarioResponseDTO.fromEntity(usuario)).build()
    }

    @POST
    @Transactional
    fun create(dto: UsuarioCreateDTO): Response {
        if (repo.existsByEmail(dto.email)) {
            return Response.status(Response.Status.CONFLICT)
                .entity(mapOf("error" to "E-mail já cadastrado"))
                .build()
        }

        val now = OffsetDateTime.now(ZoneOffset.UTC)
        val hash = passwordService.hash(dto.senha)

        val usuario = Usuario(
            id = UUID.randomUUID(),
            nome = dto.nome,
            email = dto.email,
            cargo = dto.cargo,            // int2 -> Int
            online = false,
            senhaHash = hash,
            criadoEm = now,
            atualizadoEm = now,
            lojaId = null,
            emailVerificado = false,
            twoFactorEnabled = false,
            twoFactorSecret = null,
            ultimoLoginEm = null
        )

        repo.persist(usuario)

        return Response.status(Response.Status.CREATED)
            .entity(UsuarioResponseDTO.fromEntity(usuario))
            .build()
    }
}
